src/meta/src/barrier/schedule.rs

```diff
@@ -320,63 +323,209 @@ pub struct ScheduledBarriers {
+            if db_state.checkpoint_frequency.is_none() {
+                db_state.num_uncheckpointed_barrier = 0;
+                db_state.force_checkpoint = false;
+            }
+        }
+    }
+
+    pub(super) fn update_database_barrier(
+        &mut self,
+        database_id: DatabaseId,
+        barrier_interval_ms: Option<u32>,
+        checkpoint_frequency: Option<u64>,
+    ) {
+        match self.databases.entry(database_id) {
+            Entry::Occupied(mut entry) => {
+                let db_state = entry.get_mut();
+                db_state.barrier_interval =
+                    barrier_interval_ms.map(|ms| Duration::from_millis(ms as u64));
+                db_state.checkpoint_frequency = checkpoint_frequency;
+                // Reset the `num_uncheckpointed_barrier` since the barrier interval or checkpoint frequency is changed.
+                db_state.num_uncheckpointed_barrier = 0;
+                db_state.force_checkpoint = false;
+            }
+            Entry::Vacant(entry) => {
+                entry.insert(DatabaseBarrierState::new(
+                    barrier_interval_ms,
+                    checkpoint_frequency,
+                ));
+            }
         }
+
+        // If the database already has a timer stream, reset it with the new interval.
+        let duration = if let Some(ms) = barrier_interval_ms {
+            Duration::from_millis(ms as u64)
+        } else {
+            self.sys_barrier_interval
+        };
+        self.timer_streams.insert(
+            database_id,
+            IntervalStream::new(tokio::time::interval(duration)),
+        );
     }
 
-    pub(super) fn set_min_interval(&mut self, min_interval: Duration) {
-        let set_new_interval = min_interval != self.min_interval.period();
-        if set_new_interval {
-            let mut min_interval = tokio::time::interval(min_interval);
-            min_interval.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Delay);
-            self.min_interval = min_interval;
+    /// Make the `checkpoint` of the next barrier must be true.
+    pub fn force_checkpoint_in_next_barrier(&mut self) {
+        for db_state in self.databases.values_mut() {
+            db_state.force_checkpoint = true;
         }
     }
 
     pub(super) async fn next_barrier(
         &mut self,
         context: &impl GlobalBarrierWorkerContext,
     ) -> NewBarrier {
-        let checkpoint = self.try_get_checkpoint();
-        let scheduled = select! {
+        let new_barrier = select! {
             biased;
             scheduled = context.next_scheduled() => {
-                self.min_interval.reset();
-                let checkpoint = scheduled.command.need_checkpoint() || checkpoint;
+                let database_id = scheduled.database_id;
+                // Check if the database exists.
+                assert!(self.databases.contains_key(&database_id), "database {} not found in periodic barriers", database_id);
```

Comment:
The assertion here assumes that the database should have been initialized before receiving the command. 

However, when we create the database, we don't wait to ensure that the `UpdateDatabaseBarrier` request is already handled. The `UpdateDatabaseBarrier` request is sent in a separate channel to the command channel, which means, theoretically it's possible that the command is handled earlier than the request. In this case, when we handle the command, the database is not initialized, and then we will hit this assertion.